examples/nlp/sentence_embeddings_with_sbert.py

```diff
@@ -0,0 +1,472 @@
+n is the number of sentences in the collection. For example, if n = 10000, the required
+time will be 65 hours on a V100 GPU.
+
+A common method to overcome the time overhead issue is to pass one sentence to the model,
+then average the output of the model, or take the first token (the [CLS] token) and use
+them as [sentence embedding](https://en.wikipedia.org/wiki/Sentence_embedding), then use
+a vector similarity measure like cosine similarity or Manhatten / Euclidean distance to
+find close sentences (semantically similar sentences). That will reduce the time to find
+the most similar pairs in a collection of 10,000 sentences from 65 hours to 5 seconds!
+
+If we use BERT or RoBERa directly, that will yield rather bad sentence embeddings. But if
+we fine-tune BERT using a Siamese network, that will generate semantically meaningful
+sentence embeddings. This will enable BERT to be used for new tasks. These tasks include:
+
+- Large-scale semantic similarity comparison.
+- Clustering.
+- Information retrieval via semantic search.
+
+In this example, we will show how to fine-tune a BERT or RoBERTa model using a Siamese
+network such that it will be able to produce semantically meaningful sentence embeddings
+and use them in a semantic search and clustering example.
+This method of fine-tuning was introduced in Sentence-BERT
+([SBERT](https://arxiv.org/abs/1908.10084))
+"""
+
+"""
+## Setup
+"""
+
+"""shell
+!pip install keras-nlp -q
+"""
+
+import keras_nlp
+import tensorflow as tf
+import tensorflow_datasets as tfds
+from tensorflow import keras
+import sklearn.cluster as cluster
+
+policy = keras.mixed_precision.Policy("mixed_float16")
+keras.mixed_precision.set_global_policy(policy)
+
+"""
+## Fine-tune the model using siamese networks
+"""
+
+"""
+[Siamese network](https://en.wikipedia.org/wiki/Siamese_neural_network) is a neural
+network architecture that contains two or more subnetworks. and the subnetworks share the
+same weights. It is used to generate feature vectors for each input and then compare them
+for similarity.
+
+For our example, the subnetwork will be a RoBERTa model that has a pooling layer on top
+of it to produce the embeddings of the input sentences. These embeddings will then be
+compared to each other to learn to produce semantically meaningful embeddings.
+
+The pooling strategies used are mean, max, and cls pooling. Mean pooling produces the
+best results. We will use it in our examples.
+"""
+
+"""
+### Fine-tune using regression objective function
+For building the siamese network with the regression objective function, the siamese
+network is asked to predict the cosine similarity between the embeddings of the two input
+sentences.
+
+Cosine similarity indicates the angle between the sentence embeddings. If the cosine
+similarity is high, that means there is a small angle between the embeddings; hence, they
+are semantically similar. and vice-verse.
```

Comment:
delete "and vice-verse."

In general avoid starting sentences with lowercase "and" probably